File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamunifiedmessages/UnifiedService.java
package in.dragonbra.javasteam.steam.handlers.steamunifiedmessages;

import com.google.protobuf.GeneratedMessageV3;
import in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse;
import in.dragonbra.javasteam.types.AsyncJobSingle;

/**
 * @author Lossy
 * @since 2023-01-04
 */
@SuppressWarnings("unused")
public abstract class UnifiedService {

    // private static final Logger logger = LogManager.getLogger(UnifiedService.class);

    private final SteamUnifiedMessages steamUnifiedMessages;

    public UnifiedService(SteamUnifiedMessages steamUnifiedMessages) {
        this.steamUnifiedMessages = steamUnifiedMessages;
    }

    public String getClassName() {
        return this.getClass().getSimpleName();
    }

    /**
     * @param parentClassName The parent class name, ie: Player
     * @param methodName      The calling method name, ie: GetGameBadgeLevels
     * @return The name of the RPC endpoint as formatted ServiceName.RpcName. ie: Player.GetGameBadgeLevels#1
     */
    private static String getRpcEndpoint(String parentClassName, String methodName) {
        return String.format("%s.%s#%s", parentClassName, methodName, 1);
    }

    /**
     * Sends a message.
     * <p>
     * Results are returned in a
     * {@link in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse}.
     *
     * @param message The message to send.
     * @return The JobID of the message. This can be used to find the appropriate
     * {@link in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse}.
     */
    public AsyncJobSingle<ServiceMethodResponse> sendMessage(GeneratedMessageV3 message, String methodName) {
        String serviceName = getClassName();
        String rpcEndpoint = getRpcEndpoint(serviceName, methodName);

        return sendMessageOrNotification(rpcEndpoint, message, false);
    }

    /**
     * Sends a notification.
     *
     * @param message The message to send.
     */
    public void sendNotification(GeneratedMessageV3 message, String methodName) {
        String serviceName = getClassName();
        String rpcEndpoint = getRpcEndpoint(serviceName, methodName);

        sendMessageOrNotification(rpcEndpoint, message, true);
    }

    private AsyncJobSingle<ServiceMethodResponse> sendMessageOrNotification(String rpcName, GeneratedMessageV3 message, Boolean isNotification) {

        if (isNotification) {
            steamUnifiedMessages.sendNotification(rpcName, message);
            return null;
        }

        return steamUnifiedMessages.sendMessage(rpcName, message);
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamunifiedmessages/UnifiedService.kt
package `in`.dragonbra.javasteam.steam.handlers.steamunifiedmessages

import com.google.protobuf.AbstractMessage
import com.google.protobuf.GeneratedMessage
import `in`.dragonbra.javasteam.base.PacketClientMsgProtobuf

/**
 * @author Lossy
 * @since 2024-10-22
 *
 * @constructor Abstract definition of a steam unified messages service.
 * @property unifiedMessages A reference to the [SteamUnifiedMessages] instance this service was created from.
 */
@Suppress("unused")
abstract class UnifiedService(val unifiedMessages: SteamUnifiedMessages? = null) {

    /**
     * Handles a response message for this service. This should not be called directly.
     * @param methodName The name of the method the service should handle.
     * @param packetMsg The packet message that contains the data.
     */
    abstract fun handleResponseMsg(methodName: String, packetMsg: PacketClientMsgProtobuf)

    /**
     * Handles a notification message for this service. This should not be called directly.
     * @param methodName The name of the method the service should handle.
     * @param packetMsg The packet message that contains the data.
     */
    abstract fun handleNotificationMsg(methodName: String, packetMsg: PacketClientMsgProtobuf)

    /**
     * Dispatches the provided data as a service method response.
     * @param TResponse The type of the response.
     * @param serviceClass The proto class of the response
     * @param packetMsg The packet message that contains the data.
     */
    protected fun <TResponse : GeneratedMessage.Builder<TResponse>> postResponseMsg(
        serviceClass: Class<out AbstractMessage>,
        packetMsg: PacketClientMsgProtobuf,
    ) {
        unifiedMessages?.handleResponseMsg(serviceClass, packetMsg)
    }

    /**
     * Dispatches the provided data as a service method notification.
     * @param TNotification The type of the notification.
     * @param serviceClass The proto class of the notification
     * @param packetMsg The packet message that contains the data.
     */
    protected fun <TNotification : GeneratedMessage.Builder<TNotification>> postNotificationMsg(
        serviceClass: Class<out AbstractMessage>,
        packetMsg: PacketClientMsgProtobuf,
    ) {
        unifiedMessages?.handleNotificationMsg(serviceClass, packetMsg)
    }

    /**
     * The name of the steam unified messages service.
     */
    abstract val serviceName: String
}
